[Sema] Keep attribute lists in the order the attributes were parsed

hnrklssn opened the pull request:
This renames some attribute list related functions, to make callers think about whether they want to append or prepend to the list, instead of defaulting to prepending which is often not the desired behaviour (for the cases where it matters, sometimes we're just adding to an empty list). Then it adjusts some of these calls to append where they were previously prepending. This has the effect of making `err_attributes_are_not_compatible` consistent in emitting diagnostics as `<new-attr> and <existing-attr> are not compatible`, regardless of the syntax used to apply the attributes.

hnrklssn committed:
[attrs] Rename allAll, takeAllFrom etc. for clarity (NFC)

This renames functions that prepend or append lists of attributes, to be
explicit wrt whether they prepend or append. This should make it easier
to spot mistakes, and no longer makes the prepend option seem like the
default.

hnrklssn committed:
[sema] Append attributes parsed later, instead of prepending

To make sure attributes occur the order they are parsed (for consistent
diagnostics), the most recently parsed attributes should be appended
rather than prepended.

hnrklssn committed:
format

compnerd commented on clang/include/clang/Sema/DeclSpec.h:
```diff
@@ -2638,7 +2638,7 @@ class Declarator {
   }
 
-  /// takeAttributes - Takes attributes from the given parsed-attributes
+  /// takeAttributesPrepend - Takes attributes from the given parsed-attributes
```
This is mismatched with the function

hnrklssn committed:
update doc comment

hnrklssn committed:
Rename attribute prepend/append functions (NFC)

addAll[Prepend/Append] -> [prepend/append]All
takeAllFrom[Prepend/Append] -> takeAll[Prepend/Append]From
takeAttributesFromAppend -> takeAttributesAppendFrom

hnrklssn committed:
Rename functions again (NFC)

Rename from AppendFrom/PrependFrom to AppendingFrom/PrependingFrom.

hnrklssn committed:
format

hnrklssn committed:
Rename functions (NFC)

takeAttributesAppend -> takeAttributesAppending
prependAll -> prepend
appendAll -> append

compnerd reviewed (approved):
I do still think that the altered phrasing would be a much better diagnostic, but this change now feels like it is actually cleaning up the code base a good deal. Sorry about the multiple iterations, I do appreciate the final result.

hnrklssn committed:
parsed-attributes -> ParsedAttributes
